### ISSUE TITLE
Errors occur in LrUpdaterHook when multiple optimizers are introduced

### ISSUE THREAD

**Issue by wuyuuu** [open-mmlab/mmcv#887]
Hi,

I'm using mmpose with multiple optimizers and I got errors in LrUpdateHook

```
Traceback (most recent call last):
  File "/home2/wuy/.pycharm_helpers/pydev/pydevd.py", line 1477, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home2/wuy/.pycharm_helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home2/wuy/project/mmpose/tools/model_search.py", line 169, in <module>
    main()
  File "/home2/wuy/project/mmpose/tools/model_search.py", line 165, in main
    meta=meta)
  File "/home2/wuy/project/mmpose/mmpose/apis/model_search.py", line 151, in search_model
    runner.run(data_loaders, cfg.workflow, cfg.total_epochs)
  File "/home2/wuy/project/mmpose/mmpose/apis/my_epoch_runner.py", line 134, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/home2/wuy/project/mmpose/mmpose/apis/my_epoch_runner.py", line 51, in train
    self.call_hook('before_train_iter')
  File "/home2/wuy/anaconda3/lib/python3.7/site-packages/mmcv/runner/base_runner.py", line 308, in call_hook
    getattr(hook, fn_name)(self)
  File "/home2/wuy/anaconda3/lib/python3.7/site-packages/mmcv/runner/hooks/lr_updater.py", line 139, in before_train_iter
    warmup_lr = self.get_warmup_lr(cur_iter)
  File "/home2/wuy/anaconda3/lib/python3.7/site-packages/mmcv/runner/hooks/lr_updater.py", line 88, in get_warmup_lr
    warmup_lr = [_lr * (1 - k) for _lr in self.regular_lr]
  File "/home2/wuy/anaconda3/lib/python3.7/site-packages/mmcv/runner/hooks/lr_updater.py", line 88, in <listcomp>
    warmup_lr = [_lr * (1 - k) for _lr in self.regular_lr]
TypeError: can't multiply sequence by non-int of type 'float'
```

It seems to me it is because get_warmup_lr does not work for multiple optimizers as self.regular_lr is a list for single optimizer while it is a dict for multiple optimizers.
https://github.com/open-mmlab/mmcv/blob/73bff4ead1cc3997289769c931d8c49056bbbde1/mmcv/runner/hooks/lr_updater.py#L83-L92

https://github.com/open-mmlab/mmcv/blob/73bff4ead1cc3997289769c931d8c49056bbbde1/mmcv/runner/hooks/lr_updater.py#L69-L81

**Comment by zhouzaida** [comment]
please show the optimizer which you define.

**Comment by wuyuuu** [comment]
> please show the optimizer which you define.
> I guess _lr may be str.

Thanks for your prompt reply, your guess is correct.
my optimizer is as the form: dict('model1': torch.optim.Optimizer, 'model2': torch.optim.Optimizer)
```
{'model': SGD (
Parameter Group 0
    dampening: 0
    lr: 0.001
    momentum: 0
    nesterov: False
    weight_decay: 0
), 'model2': SGD (
Parameter Group 0
    dampening: 0
    lr: 0.001
    momentum: 0
    nesterov: False
    weight_decay: 0
)}
```
BTW, self.regular_lr is `{'model': [0.001], 'model2': [0.001]}`
and _lr becomes `'model1'` and `'model2'`

**Comment by zhouzaida** [comment]
Should `MomentumUpdaterHook `also support multiple optimizers like the `LRUpdaterHook`?

https://github.com/open-mmlab/mmcv/blob/371a21728fdb508f7422942267af408db3503e2f/mmcv/runner/hooks/momentum_updater.py#L44-L48

https://github.com/open-mmlab/mmcv/blob/371a21728fdb508f7422942267af408db3503e2f/mmcv/runner/hooks/lr_updater.py#L69-L81

**Comment by wuyuuu** [comment]
> Should `MomentumUpdaterHook `also support multiple optimizers like the `LRUpdaterHook`?
> 
> https://github.com/open-mmlab/mmcv/blob/371a21728fdb508f7422942267af408db3503e2f/mmcv/runner/hooks/momentum_updater.py#L44-L48
> 
> https://github.com/open-mmlab/mmcv/blob/371a21728fdb508f7422942267af408db3503e2f/mmcv/runner/hooks/lr_updater.py#L69-L81

Hello Zhou,

To my best knowledge, I think the MomentumHook should set different momentums for different optimizers. ( Now I'm using Adam optimizer for my project which incorporates the momentum and I think surely different Adam optimizers should include different momentums)

BTW, I simply modified the LrUpdateHook by: 
```
        def get_warmup_lr(self, cur_iters):
        if isinstance(self.regular_lr, dict):
            w = [_lr[0] for _lr in self.regular_lr.values()]
        else:
            w = self.regular_lr
        if self.warmup == 'constant':
            warmup_lr = [_lr * self.warmup_ratio for _lr in w]
        elif self.warmup == 'linear':
            k = (1 - cur_iters / self.warmup_iters) * (1 - self.warmup_ratio)
            warmup_lr = [_lr * (1 - k) for _lr in w]
        elif self.warmup == 'exp':
            k = self.warmup_ratio ** (1 - cur_iters / self.warmup_iters)
            warmup_lr = [_lr * k for _lr in w]
        if isinstance(self.regular_lr, dict):
            return {k: [_lr] for k, _lr in zip(self.regular_lr.keys(), warmup_lr)}
        return warmup_lr
```
So far it works well for me.

**Comment by zhouzaida** [comment]
> > Should `MomentumUpdaterHook `also support multiple optimizers like the `LRUpdaterHook`?
> > https://github.com/open-mmlab/mmcv/blob/371a21728fdb508f7422942267af408db3503e2f/mmcv/runner/hooks/momentum_updater.py#L44-L48
> > 
> > https://github.com/open-mmlab/mmcv/blob/371a21728fdb508f7422942267af408db3503e2f/mmcv/runner/hooks/lr_updater.py#L69-L81
> 
> Hello Zhou,
> 
> To my best knowledge, I think the MomentumHook should set different momentums for different optimizers. ( Now I'm using Adam optimizer for my project which incorporates the momentum and I think surely different Adam optimizers should include different momentums)
> 
> BTW, I simply modified the LrUpdateHook by:
> 
> ```
>         def get_warmup_lr(self, cur_iters):
>         if isinstance(self.regular_lr, dict):
>             w = [_lr[0] for _lr in self.regular_lr.values()]
>         else:
>             w = self.regular_lr
>         if self.warmup == 'constant':
>             warmup_lr = [_lr * self.warmup_ratio for _lr in w]
>         elif self.warmup == 'linear':
>             k = (1 - cur_iters / self.warmup_iters) * (1 - self.warmup_ratio)
>             warmup_lr = [_lr * (1 - k) for _lr in w]
>         elif self.warmup == 'exp':
>             k = self.warmup_ratio ** (1 - cur_iters / self.warmup_iters)
>             warmup_lr = [_lr * k for _lr in w]
>         if isinstance(self.regular_lr, dict):
>             return {k: [_lr] for k, _lr in zip(self.regular_lr.keys(), warmup_lr)}
>         return warmup_lr
> ```
> 
> So far it works well for me.

Thanks for your feedback, we will support what you suggest in the next version.

**Comment by zhouzaida** [comment]
Related PR https://github.com/open-mmlab/mmcv/pull/907

**Comment by zhouzaida** [comment]
hi, @wuyuuu, the issue is resolved by pr #907 merged into master.
